Design Document for Filter API

georgi-l95 opened the pull request:
**Description**:
Adds new design doc for introduction of Filter API, alongside couple of new methods - `eth_newFilter`, `eth_uninstallFilter`, `eth_getFilterChanges` and `eth_getFilterLogs`

**Related issue(s)**:

Fixes #1484 

**Notes for reviewer**:
<!-- Provide logs, performance numbers or screenshots of the new functionality -->

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

Ivo-Yankov commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,133 @@
+
+#### Parameters
+
+An existing
```
```suggestion
An existing FILTER_ID
```

georgi-l95 replied:
Thanks, added.

Nana-EC commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,133 @@
+
+|      Filter Type       |                                                    Description                                                     |                           Support                           |
+| :--------------------: | :----------------------------------------------------------------------------------------------------------------: | :---------------------------------------------------------: |
+|          logs          |                               Returns all logs matching filter with given filter ID.                               |                       Yes, initially                        |
```
Can you speak to the time frame of the logs. I believe it's all the logs since the last query to the filter correct?

georgi-l95 replied:
Hm, we should follow `eth_getLogs` limitations. `eth_getFilterChanges` will give you all new logs since last poll and `eth_getFilterLogs` gives you all logs for the provided filters.

Nana-EC commented on docs/design/eth-filter.md:
````diff
@@ -0,0 +1,133 @@
+#### Request
+
+```javascript
+{"jsonrpc":"2.0","id": 1, "method": "eth_newFilter", "params": [fromBlock, toBlock, address, topics]}
````
Q's
1. Can a `toBlock` be in the future?
2. If a `toBlock` is in the past is this call intended to give you all the logs in one go?
3. Are there limits to the amount of logs we should return in either case

georgi-l95 replied:
1. Should be able to set in future.
2. Yes, if they don't exceed the predefined limit. (similar to the one we use in `eth_getLogs`)
3. I think 2 answer this.

Nana-EC commented on docs/design/eth-filter.md:
````diff
@@ -0,0 +1,133 @@
+#### Response
+
+```javascript
+{"jsonrpc":"2.0","id": 1, "result": FILTER_ID}
````
Q: what format will the FILTER_ID take? Is it an integer, a hex string ...

georgi-l95 replied:
Sorry, added, it should be hex.

Nana-EC commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,133 @@
+
+| Error Codes |   Error message   |                              Solution                               |
+| :---------: | :---------------: | :-----------------------------------------------------------------: |
+|    32000    | Filter not found. | Occurs when user attempts receive logs using non-existing filterID. |
```
Q: 
1. are there any error codes around range?
2. What happens if during a request Mirror Node fails to give us logs, how do we intend to behave and respond?

georgi-l95 replied:
1. I've copied the error codes that other providers are using.
2. Good questions, we should either our defined rpc error ,or empty array(this is what other rpc providers are returning if there are no logs to return)

Nana-EC commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,133 @@
+
+## Rate Limits
+
+1. All filters expire after 5 minutes of inactivity (no queries).
```
This should be configureable. Suggest a config label

georgi-l95 replied:
Added.

Nana-EC commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,133 @@
+
+## Rate Limits
+
+1. All filters expire after 5 minutes of inactivity (no queries).
```
Q: What about filter count limits? Is there a max number of filters we should support? If so this should be configureable also.

georgi-l95 replied:
Good questions, not sure if we should start with limit or wait and see how many people are using it and if there is performance impact.

Nana-EC replied:
Start. Harder to add after and limit. So we should set a limit like we did for `eth_subscribe` and have errors for that.

georgi-l95 replied:
Okay, added.

Nana-EC commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,133 @@
+
+Capture metrics for the following:
+
+1. Log every call to all filter API method, as a total amount and broken down by IP.
```
Q's
1. What purpose is IP serving?
3. What about number of filters

AlfredoG87 replied:
Yeah I agree that we should remove unique values labels for metrics since they drastically increase cardinality and costs.

georgi-l95 replied:
Gotcha, removed.

Nana-EC commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,133 @@
+The following test cases should be covered but additional tests would be welcome.
+
+1. Overall functionality of creating and uninstalling new filters.
+2. Receiving requested information depending on filterID using `eth_getFilterChanges` and `eth_getFilterLogs`.
```
Under this we'd want
1. Case where no logs applicable
2. Case where there are logs within range
3. Case where there's a logs out of range situation

georgi-l95 replied:
Added.

Nana-EC commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,133 @@
+
+## Non-Functional Requirements
+
+Users should be required to renew their filters in the case of an inactivity. If for some reason the relay is restarted it will lose all cached filters.
```
Q: How would this work?
Do they not just call `eth_newFilter` again?

georgi-l95 replied:
Yes, they need to call eth_newFilter again with their filter params. Added to the doc.

Nana-EC commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,133 @@
+
+## Answered Questions
+
+1. How will we implement the `getFilterChanges` and `getFilterLogs` flow?
```
I'm not sure 1 & 2 are fully answered.
As part of the design doc I'd like to get a better sense of what classes are involved, both existing and new ones.
These changes should be called out in summary

georgi-l95 replied:
Yes, you are right. I've reworked most of the doc.

Nana-EC replied:
I don't think you've answered the flow or interfaces much.
This isn't new logic i guess since it's similar to `eth_getLogs`

Nana-EC commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,133 @@
+|      Filter Type       |                                                    Description                                                     |                           Support                           |
+| :--------------------: | :----------------------------------------------------------------------------------------------------------------: | :---------------------------------------------------------: |
+|          logs          |                               Returns all logs matching filter with given filter ID.                               |                       Yes, initially                        |
+|       newBlocks        |                            Returns all new blocks matching filter with given filter ID.                            |                 No, maybe in future release                 |
```
Q: Do we know how often this is utilized in other RPCs?

georgi-l95 replied:
Didn't find any information regarding this.

Nana-EC reviewed (changes requested):
Great start.
Lots of questions and suggestions 😄

georgi-l95 commented:
> Thank you @georgi-l95. Some general thoughts. I know this isn't a HIP and I'm not sure where exactly we draw the line between HIP and Design Doc, but I think it will be useful for anyone reviewing this document to have a bit more information:
> 
> 1. Several example user stories that explain _how_ the Filter API is used and _why_ it's useful. Code snippet examples will help clarify things as well.
> 2. Exactly what params each endpoint supports and what they signify. For example, where are the Filter Types used if the `eth_newFilter` method doesn't accept a `type` parameter?
> 3. What are the answers to the `Answered Questions` listed below?
> 4. If I understand correctly, once a user creates a Filter, that Filter is "living" on the Relay side and the user can reference it by its Filter ID. As @Ivo-Yankov pointed out in a separate channel, we should briefly describe the current mainnet Relay high-level architecture (how many Relay servers are operational, are they load-balanced, do they share a cache) and design the filters to be "shareable" between different Relay instances if possible, we should even try to be future-proof about this. Is this the first stateful feature we're implementing on the Relay side?

Thank you for the review.
1. I've added user stories to the HIP. Should I add them also here ? 
2. Added needed params for each method and how they affect each other.
3. Reworked some of the answered questions. Basically these questions are answered in the doc itself.
4. Yes, for us to be able to support this new functionally properly, we'd need to add shared cache across instances. I think eth_subscribe is our full stateful method.

georgi-l95 commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,133 @@
+
+## Purpose
+
+The JSON-RPC Relay currently doesn't support monitoring events using filter API as is common in many other relays and is expected by many web3 tools, which makes polling for event logs, new blocks and transactions hard for developers.
```
I've added the ones that 100% use it.

AlfredoG87 commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,201 @@
+3. The duration of active filters.
+4. The total amount of requests to the mirror node per filter.
+5. The amount of requests to the mirror node per filter that have returned non-null data.
+
```
6. Amount of filters that were uninstalled by Time Limit vs Explicitly by the user.

georgi-l95 replied:
Thanks, added.

AlfredoG87 commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,201 @@
+
+## Non-Functional Requirements
+
+Users should be required to renew their filters, by calling again `eth_newFilter`, in the case of an inactivity. If for some reason the relay is restarted it will lose all cached filters.
```
should time limits restart every time a filter is used within `eth_getFilterChanges` or `eth_getFilterLogs`, since is sign of activity?

georgi-l95 replied:
Yes, filter should be deleted only if it's not used in the last 5 minutes (for example) or whatever our configuration of this property is.

georgi-l95 commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,201 @@
+5. Case where logs are not within limit range.
+6. Case where no logs are available.
+7. E2E test using popular libraries (`ethers.js WebSocketProvider`).
+
```
Yes, added those test cases, thank you.

georgi-l95 commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,201 @@
+
+## Non-Functional Requirements
+
+Users should be required to renew their filters, by calling again `eth_newFilter`, in the case of an inactivity. If for some reason the relay is restarted it will lose all cached filters.
```
Yes, you're right, in the case of distributed cache, in normal scenarios it will only be deleted by expired TTL.

georgi-l95 commented on docs/design/eth-filter.md:
```diff
@@ -0,0 +1,201 @@
+
+## Deployment
+
+Filter API will run alongside the already available HTTP server.
```
Yep, will run on the same http server.

Nana-EC reviewed (approved):
LG, updates should be made as implementation alters items
